rust_build_utils/msvc.py

```diff
@@ -0,0 +1,149 @@
+import subprocess
+from contextlib import contextmanager
+from pathlib import Path
+import re
+import os
+from typing import Optional
+
+
+@contextmanager
+def activate_msvc(
+    arch: str,
+    version_preference: Optional[str] = None,
+    edition_preference: Optional[str] = None,
+    direct_pass_arch: bool = False,
+):
+    """Activate MSVC tools for building a specific arch
+
+    Arguments:
+    arch: build output (target) architecture. The arch will be appended to 'amd64_'
+      (because our hosts are x64; if x64 or amd64 is given it will be passed directly) and passed to vcvarsall.bat.
+      aarch64 will be converted to arm64
+      For example:
+        amd64   -> vcvarsall.bat amd64
+        x64     -> vcvarsall.bat x64
+        x86     -> vcvarsall.bat amd64_x86
+        arm     -> vcvarsall.bat amd64_arm
+        arm64   -> vcvarsall.bat amd64_arm64
+        aarch64 -> vcvarsall.bat amd64_arm64
+      https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170#vcvarsall-syntax
+    version_preference (optional): version (year) preference (for example: "2022").
+      When requested version is not found an exception will be raised. If not given, a highest version will be used.
+      Folders in 'C:\Program Files\Microsoft Visual Studio' can be used as versions.
+    version_preference (optional): edition preference. There can be multiple editions of VS installed at the time.
+      If requested edition is not found in automatically (or explicitly) chosen version,
+      an exception will be raised. Example values are Community, Professional, Enterprise, BuildTools, Preview.
+      If not given, a preference is given (in same order): BuildTools, Enterprise, Professional, Community, Preview.
+      If none of these is found, edition will be picked at random.
+      The editions are folders in 'C:\Program Files\Microsoft Visual Studio\<version>'.
+    direct_pass_arch (optional): if True, the `arch` value will be passed to vcvarsall.bat
+      without appending the 'amd64_' prefix.
+
+    Example usage:
+    with activate_msvc('arm64'):
+        print(subprocess.run("link"))
+        print(subprocess.run("cl"))
+    """
+
+    # Sample location of vcvarsall script:
+    # "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat"
+    # We begin by finding microsoft visual studio installation.
+    mvs_path = Path(r"C:\Program Files\Microsoft Visual Studio")

```

Comment:
The default path in my experience for Visual Studio is "C:\Program Files (x86)\Microsoft Visual Studio". I assume you've tested it on some machine where it was "Program Files" instead, so I guess it's worth checking both options

Reply:
> The default path in my experience for Visual Studio is "C:\Program Files (x86)\Microsoft Visual Studio". I assume you've tested it on some machine where it was "Program Files" instead, so I guess it's worth checking both options

Not for the scripts: https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170#developer_command_file_locations
But to allow full configuration a version of the function with explicit path should probably be provided as well

Reply:
It's most likely a bug in their documentation :) Few lines below that they have this:
```
Create your own command prompt shortcut
Open the Properties dialog for a developer command prompt shortcut to see the command target used. For example, the target for the x64 Native Tools Command Prompt for VS 2019 shortcut is something similar to:

%comspec% /k "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars64.bat"
```
But  for example on GitHub runners it's "Program Files" so in general those two should be checked

Reply:
Yes I will add a check for both. I guess this is very much version dependent, because 2022 version is installed in `Program Files` on our runners :shrug: